src/dotty/tools/dotc/transform/Space.scala

```diff
@@ -0,0 +1,316 @@
+      }.toList
+
+      debug(s"child of ${t.show}: ${children.map(_.show).mkString(", ")}")
+
+      children
+  }
+
+  // parent is specialized, specialize child as well
+  def specializeChild(p: Type, c: Type): Type = p match {
+    case p: RefinedType => p.wrapIfMember(specializeChild(p.parent, c))
+    case _ => c
+  }
+
+  // abstract sealed types and or types can be decomposed
+  def canDecompose(t: Type): Boolean = {
+    t.typeSymbol.is(Flags.allOf(Flags.Abstract, Flags.Sealed)) || t.isInstanceOf[OrType]
+  }
+
+  def isCaseClass(t: Type): Boolean = t.classSymbol.isClass && t.classSymbol.is(Flags.CaseClass)
+
+  // display spaces
+  def show(s: Space): String = s match {
+    case Empty => ""
+    case Typ(t) =>
+      val sym = t.widen.classSymbol
+
+      if (sym is Flags.ModuleClass)
+        sym.sourceModule.name.show
+      else if (t.isInstanceOf[OrType])
+        "_: " + t.show
+      else if (getSignature(t).isEmpty)
+        "_: " + t.classSymbol.name
+      else if (ctx.definitions.isTupleType(t))
+        "(" + getSignature(t).map(_ => "_").mkString(", ") + ")"
+      else
+        t.classSymbol.name + "(" + getSignature(t).map(_ => "_").mkString(",") + ")"
+    case Kon(t, params) =>
+      if (ctx.definitions.isTupleType(t))
+        "(" + params.map(show _).mkString(", ") + ")"
+      else
+        t.classSymbol.name + "(" + params.map(show _).mkString(", ") + ")"
+    case Or(spaces) => spaces.map(show _).mkString(", ")
+  }
+
+  // does the type t have @unchecked annotation
+  def skipCheck(t: Type): Boolean = t match {
+    case AnnotatedType(t, annot) => ctx.definitions.UncheckedAnnot == annot.symbol
+    case _ => false
+  }
+
+  // widen a type and eliminate anonymous class
+  private def widen(t: Type): Type = t.widen match {
+    case t:TypeRef if t.symbol.isAnonymousClass =>
+      t.symbol.asClass.typeRef.asSeenFrom(t.prefix, t.symbol.owner)
+    case t => t
+  }
+
+  def exhaustivity(_match: tpd.Match): Unit = {
+    val Match(sel, cases) = _match
+    val selTyp = widen(sel.tpe)
+
+    debug(s"====patterns:\n${cases.map(_.pat).mkString("\n")}")
+    debug(s"====pattern space:\n${show(cases.map(x => project(x.pat)).reduce((a, b) => Or(List(a, b))))}")
+    val patternSpace = simplify(cases.map(x => project(x.pat)).reduce((a, b) => Or(List(a, b))))
+    debug(s"====selector:\n" + selTyp)
+    debug("====pattern space:\n" + show(patternSpace))
+    val uncovered = simplify(minus(Typ(selTyp), patternSpace))
+
+    if (uncovered != Empty) {
+      ctx.warning(
+        "match may not be exhaustive.\n" +
+        "It would fail on the following inputs: " +
+        show(uncovered), _match.pos
+      )
+    }
+  }
+
+  // find the first redundant pattern and stop
+  def redundancy(_match: tpd.Match): Unit = {
+    val Match(sel, cases) = _match
+    val selTyp = widen(sel.tpe)
+
+    // starts from the second, the first can't be redundant
+    var i = 1
+    while(i < cases.length) {
```

Comment:
The cost depends on the program, but for normal programs the cost should be negligible. I'll do as you suggested. Thanks.